File: backend/app/schemas/gyne_diagnosis.py
from pydantic import BaseModel, ConfigDict, field_validator
from datetime import datetime
from typing import Optional, List


# --- Master Data Schemas ---

class GyneSpecimenAdequacyBase(BaseModel):
    id: int
    group_type: str
    text: str
    code: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class GyneDiagnosisCategoryBase(BaseModel):
    id: int
    code: str
    text: str
    parent_id: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


# --- Diagnosis Schemas ---

class GyneSignerSchema(BaseModel):
    user_id: int
    role: str = "primary"
    signed_at: Optional[datetime] = None


class GyneDiagnosisBase(BaseModel):
    # Old text fields (Keep optional or deprecate)
    adequacy: Optional[str] = None
    category: Optional[str] = None
    interpretation: Optional[str] = None
    note: Optional[str] = None
    revised_reason: Optional[str] = None

    # New structured fields
    adequacy_id: Optional[int] = None
    endocervical_status_id: Optional[int] = None
    quality_id: Optional[int] = None
    category_1_id: Optional[int] = None
    category_2_id: Optional[int] = None
    
    signers: Optional[List[GyneSignerSchema]] = []


class GyneDiagnosisCreate(GyneDiagnosisBase):
    case_id: int


class GyneDiagnosisUpdate(BaseModel):
    # Free text fields
    adequacy: Optional[str] = None
    category: Optional[str] = None
    interpretation: Optional[str] = None
    note: Optional[str] = None
    
    # New structured fields
    adequacy_id: Optional[int] = None
    endocervical_status_id: Optional[int] = None
    quality_id: Optional[int] = None
    category_1_id: Optional[int] = None
    category_2_id: Optional[int] = None
    
    signers: Optional[List[GyneSignerSchema]] = None

    revised_reason: Optional[str] = None

    @field_validator("revised_reason")
    @classmethod
    def reason_must_not_be_empty(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            if not v or len(v.strip()) < 5:
                raise ValueError("Revised reason must be at least 5 characters long")
        return v


class GyneDiagnosisResponse(GyneDiagnosisBase):
    id: int
    case_id: int
    version: int
    is_current: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    # Relationships (Optional: include if needed for display)
    adequacy_obj: Optional[GyneSpecimenAdequacyBase] = None
    endocervical_status_obj: Optional[GyneSpecimenAdequacyBase] = None
    quality_obj: Optional[GyneSpecimenAdequacyBase] = None
    category_1_obj: Optional[GyneDiagnosisCategoryBase] = None
    category_2_obj: Optional[GyneDiagnosisCategoryBase] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/gyne-diagnosis.ts
// Master Data Interfaces
export interface GyneSpecimenAdequacy {
  id: number;
  group_type: string; // "ADEQUACY", "ZONE", "QUALITY"
  text: string;
  code?: string;
}

export interface GyneDiagnosisCategory {
  id: number;
  code: string;
  text: string;
  parent_id?: number | null;
}

// Diagnosis Interfaces
export interface GyneDiagnosisResponse {
  id: number;
  case_id: number;
  
  // Legacy / Text Fields
  adequacy: string | null;
  category: string | null; 
  interpretation: string | null;
  note: string | null;
  
  // New Structured Fields
  adequacy_id: number | null;
  endocervical_status_id: number | null;
  quality_id: number | null;
  category_1_id: number | null;
  category_2_id: number | null;

  version: number;
  is_current: boolean;
  revised_reason: string | null;
  created_at: string;
  updated_at: string | null;

  // Joined Objects (Optional)
  adequacy_obj?: GyneSpecimenAdequacy;
  endocervical_status_obj?: GyneSpecimenAdequacy;
  quality_obj?: GyneSpecimenAdequacy;
  category_1_obj?: GyneDiagnosisCategory;
  category_2_obj?: GyneDiagnosisCategory;

  signers?: { user_id: number; role: string; signed_at?: string | null }[];
}

export interface GyneDiagnosisCreate {
  case_id: number;
  
  // Text fields (Optional now)
  adequacy?: string;
  category?: string;
  interpretation?: string;
  note?: string;

  // New Structured Fields
  adequacy_id?: number;
  endocervical_status_id?: number;
  quality_id?: number;
  category_1_id?: number;
  category_2_id?: number;

  signers?: { user_id: number; role: string; signed_at?: string | null }[];
}

export interface GyneDiagnosisUpdate {
  // Text fields
  adequacy?: string;
  category?: string;
  interpretation?: string;
  note?: string;

  // New Structured Fields
  adequacy_id?: number;
  endocervical_status_id?: number;
  quality_id?: number;
  category_1_id?: number;
  category_2_id?: number;

  signers?: { user_id: number; role: string; signed_at?: string | null }[];

  revised_reason?: string;
  slide_quality?: string;
  stain_quality?: string;
}
